### ISSUE TITLE
We should switch to a pyproject.toml

### ISSUE THREAD

**Issue by john-science** [terrapower/armi#1124]
Two important things have happened in the Python world which break our current build system:

* [The Python ORG team has officially removed `distutils` from the Python language.](https://peps.python.org/pep-0632/)
* [The PIP team has decided to significantly reduce the powers we use in `requirements.txt`.](https://github.com/pypa/pip/issues/8210)

As such, I want to investigate moving ARMI to [pyproject.toml](https://packaging.python.org/en/latest/tutorials/packaging-projects/#creating-pyproject-toml). 

Such a decision would affect our user base, and I don't want to surprise anyone.  But the tools we are using to build ARMI are deprecating or straight deprecated, so the change might be mandatory.

**Comment by john-science** [comment]
@opotowsky @sombrereau @jakehader @drewj-usnctech  You are my major stakeholders, and I just want you to be aware of this.

**Comment by opotowsky** [comment]
I'm on board with whatever needs to be done. No one wants github actions tapping out on us again!

**Comment by drewj-usnctech** [comment]
Big fan of `pyproject.toml` files. We just use something like `pip install /path/to/armi` so any build changes would be largely transparent to us.

**Comment by john-science** [comment]
I just need to save [a link](https://github.com/terrapower/armi/actions/runs/4009943982/jobs/6885882968) to the original error message, just in case:

```
test: install_package_deps> python -I -m pip install configparser coverage<=6.5.0 future h5py>=3.0 htmltree
matplotlib numpy<=1.23.5,>=1.21 ordered-set pillow pluggy pyDOE pyevtk scipy tabulate voluptuous xlrd
yamlize -cD:\a\armi\armi\.tox\test\constraints.txt

DEPRECATION: Constraints are only allowed to take the form of a package name and a version specifier.
Other forms were originally permitted as an accident of the implementation, but were undocumented.
The new implementation of the resolver no longer supports these forms. A possible replacement is
replacing the constraint with a requirement.

Discussion can be found at https://github.com/pypa/pip/issues/8210
```

**Comment by ntouran** [comment]
The closer we can get to the current standard recommended packaging, the better IMHO, if it doesn't cause undo pain. 

https://packaging.python.org/en/latest/

**Comment by john-science** [comment]
I tried `pyproject.toml` in a local ARMI repo, and it broadly works.

I haven't tested it heavily in our downstream plugin ecosystem yet. But there will be potentially be a couple of changes to the steps people take to install their code locally. Mostly, I'm just trying to make sure I can reduce/eliminate the number of these breaking changes. Not all of our user base is really interested in software, so I want to make it painless.

Though there are a couple of features and use-cases we have that are broken (or changed enough to be awkward), so I'm working on those:

1. Currently, we store our version in code, and parse it in a `setup.py` using code. It appears that is impossible in `pyproject.toml`, and they now suggest storing the version in `pyproject.toml` and reading it in code using `import toml.`  Which is... fine. But it will be a change I have to make in like 65 different repos... boring!
2. Our users currently install ARMI doing `pip install -r requirements.txt`, but that line/option has been removed from `requirements.txt`, so that means our user base might potentially have to type one more line in to install all their codebases.  There are options here, I'm trying to find a good one.
3. We package a LOT of little data files with ARMI. The new solutions for this in `pyproject.toml` seem ugly and will requirement too much boilerplate to implement as well. Maybe I know the solution for this, but I just don't like it.

I am hopeful that my slow, contemplative approach here is fine because I have all the time in the world. Fingers crossed.

**Comment by opotowsky** [comment]
For number 1, I think there are a lot of cases where the version is stored in `setup.py`, actually. So this hopefully isn't as much work as you're thinking. 

For number 2, what is the new restriction? Is a list of `git+<url>@main` no longer an option in the requirements files?

**Comment by ntouran** [comment]
> We package a LOT of little data files with ARMI. The new solutions for this in pyproject.toml seem ugly and will requirement too much boilerplate to implement as well. Maybe I know the solution for this, but I just don't like it.

I'd love to merge our `pylintrc` into `pyproject.toml`'s linting section so that tools like `ruff` will "just work" a little bit better. Do you have a branch pushed with your `pyproject.toml` experiments? I'd like to contribute to it

**Comment by john-science** [comment]
> I'd love to move our `pylintrc` into `pyproject.toml`'s linting section...

Please note that we can not officially include `pylint` in our repo, as it is licensed under an anti-competitive GNU license: https://pypi.org/project/pylint/

Pylint is a fine tool, but we will want to find something else to standardize on.

**Comment by ntouran** [comment]
Including pylint is a whole lot different from including pylint configuration. The latter is perfectly acceptable. https://github.com/terrapower/armi/blob/main/pylintrc

**Comment by john-science** [comment]
I have too many tabs open in my browser, and I want to save these off somewhere:

* [PyPA Best Practices](https://packaging.python.org/en/latest/tutorials/packaging-projects/#creating-pyproject-toml)
* [Official Docs](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/)
* [SO: pyproject vs setup.py](https://stackoverflow.com/questions/71193095/questions-on-pyproject-toml-vs-setup-py)
* [Transition your Python project to use pyproject.toml ](https://stackoverflow.com/questions/72018887/how-to-build-a-universal-wheel-with-pyproject-toml)
* [Cheat Sheet](https://betterprogramming.pub/a-pyproject-toml-developers-cheat-sheet-5782801fb3ed)
* [Practical Guide](https://godatadriven.com/blog/a-practical-guide-to-setuptools-and-pyproject-toml/)
* [Building a Wheel with pyproject.toml](https://textbooks.cs.ksu.edu/cc410/ii-gui/15-releases/07-building-wheel/)
* [another SO](https://stackoverflow.com/questions/64150719/how-to-write-a-minimally-working-pyproject-toml-file-that-can-install-packages/64151860#64151860)
* [relevant SO](https://stackoverflow.com/questions/58708848/how-to-build-python-project-based-on-pyproject-toml)
* [Windows Wheels](https://stackoverflow.com/questions/26059111/build-a-wheel-egg-and-all-dependencies-for-a-python-project)
* [Example: Pandas](https://github.com/pandas-dev/pandas/blob/main/pyproject.toml)

**Comment by john-science** [comment]
So, I almost have a working version of this. 

The problem is we still have a very small user base who needs us to support Python 3.7. 

And, I'm having trouble making this work for the 3.7 crowd.

Specifically, in the `pyproject.toml` world, you set the version of your app in the `pyproject.toml` file and not in the code as we do now.  And you get that version in the code by doing:

```python
import importlib.metadata
__version__ = importlib.metadata.version("armi")
```

But this `importlib.metadata` isn't [available until Python 3.8](https://pypi.org/project/importlib-metadata/#files).

@sombrereau @opotowsky Just FYI.
